### PR TITLE
Collect even more links from even more sources

### PR DESCRIPTION
**Pull request by da2x**
### Detailed description

* Support area, base, blockquote, object, and source elements.
* Support srcset attribute on img and source elements (for responsive images).
* Support poster attribute on the video element.
* Support ping attribute on a and area elements.
* Support HTML+RDFa global about and resource attributes.

### To do

* [x] A test including an image srcset attribute (I still can’t run the tests locally)

### CODE REVIEW

**Comment by ddfreyne** [comment]
For the tests: I did a bit of digging and found

- https://stackoverflow.com/questions/51661931/rails-s-command-issue/51662528
- https://github.com/cowboyd/therubyracer/issues/453

I think with `therubyracer` commented-out in the Gemfile, the tests might start to work. (Some other tests might fail, but at least it’s a start.)

**Comment by da2x** [comment]
So the mixed content filter depends on `resource_uris_in_file` only checking document embeds. I didn’t quite understand the difference between it and `hrefs_in_file` when looking at the code. OK, so all elements but `a` and `area` (unless they point to a document under the `base_uri` or the `base` element’s URI) can affect mixed content and the `GLOBAL_ATTR`s don’t either.

To get the mixed content tester working the remaining issues are:

 - [ ] Does Nokogiri handle and return absolute URIs for document with a base URI or does Nanoc needto handle that?
 - [ ] Return `a` and `area` elements linking to the base_uri in the site configuration in `resource_uris_in_file`.
 - [ ] Filter out the about and resource attributes when running the mixed content test.

I’m inclined to say this is out of scope for this PR as the mixed content checker didn’t really do it’s job too well prior to this change; and this change will let the internal and external link checkers do a way better job.

**Comment by ddfreyne** [comment]
@da2x I’m rather confused by the original code myself now. :( (I assure you that this does not happen very often.)

`LinkCollector` would also better be named `URLCollector` I believe, since it collects URLs (for things to check). It might be pedantic, but URLs ≠ links, and I think the distinction is important in this case.

**Comment by ddfreyne** [comment]
More potential pedantry: tags ≠ elements, even though the LinkCollector refers to elements as tags.

**Comment by da2x** [comment]
It’s actually a URI collector and not a URL collector to be even more pedantic as it picks up on all URI schemes.

Anyhow, I’m unsure how to proceed with this. How about dropping the mixed content checker altogether? The current version wouldn’t pick up on a bunch of stuff yet I can’t find any bugs relating to it. To me, this suggests that no one is using it.

The mixed content checker could also be changed to check all links to the site conf base_uri domain under the http scheme. This would also pick up on internal links that could downgrade security. This wouldn’t pick up on external embeds, however.

**Comment by ddfreyne** [comment]
I’d like to keep the `mixed_content` check; it’s mentioned on the web site (https://nanoc.ws/doc/testing/ near the bottom).

Would it make sense to split `LinkCollector` into one class for collecting URIs for the link checker, and another one for mixed content? The mixed-content one might still not be working great, but at least you’d have the freedom to structure `URICollector` the way you want and let it do what you need.

**Comment by da2x** [comment]
I’ll change things so `resource_uris_in_file` (for `:mixed_content`) returns embeds and `hrefs_in_file` returns everything (for `:ilinks` and `:elinks`). I’ll then fix al three checkers to use the right one. As a bonus this should speed up each check as each check will only need one pass through all the output files.

**Comment by da2x** [comment]
Current status: Everything should be working including the mixed content checker (except downgrade links but those were already not tested). The internal_link checker now handles deeply-nested relativized links and more variations of same-directory file links.

I still can’t run the tests locally even when commenting out therubyracer. I’m on Fedora so gcc hardening is applied, however. I don’t want to blindly writing more tests, however.

@ddfreyne, what is the failing test? I can’t tell from the output. The same tests failed in #1424.

I believe this should be at a mergeable state now.

**Comment by da2x** [comment]
So there was a number of issues higher than zero. I’ve addressed those now and added tests for new link collection sources, and changed the ilinks tests to work with URIs from the link collector.

I don’t know whats with `test_nanoc_cruby25` and the CCI output doesn’t give me much to work with. `test_nanoc_cruby24` times out after ten minutes but there are no details on what it got hung up on. Both also failed in the https://github.com/nanoc/nanoc/pull/1424 hotfix with the same CCI output.

**Comment by ddfreyne** [comment]
Yeah, the tests on CI started to become flaky and I have not investigated in detail yet. I plan on doing that soon!

File: nanoc/test/extra/test_link_collector.rb [hunk 1]
```diff
@@ -16,95 +19,157 @@ def test_all
...
     refute_includes hrefs, 'http://example.com/'
     refute_includes hrefs, 'https://example.com/'
-    assert_includes hrefs, 'stuff/'
+    refute_includes hrefs, 'https://nanoc.ws'
     refute_includes hrefs, 'mailto:bob@example.com'
-    assert_includes hrefs, '../stuff'
-    assert_includes hrefs, '/stuff'
+  end
+
+  def collect_links_from_space_separated_lists
```
Comment (ddfreyne) [inline nanoc/test/extra/test_link_collector.rb]: Test method names have to start with `test_`, or they won’t be run

File: nanoc/test/extra/test_link_collector.rb [hunk 2]
```diff
@@ -16,95 +19,157 @@ def test_all
...
+    assert_includes hrefs, path_to_file_uri('ping1', Dir.pwd)
+    assert_includes hrefs, path_to_file_uri('ping2', Dir.pwd)
+    assert_includes hrefs, path_to_file_uri('ping5', Dir.pwd)
+    assert_includes hrefs, path_to_file_uri('ping3', Dir.pwd)
+    assert_includes hrefs, 'http://example.com/ping4'
+    refute_includes hrefs, nil
+    refute_includes hrefs, path_to_file_uri('/', Dir.pwd)
+  end
+
+  def collects_exotic_links
```
Comment (ddfreyne) [inline nanoc/test/extra/test_link_collector.rb]: Same here: add `test_` prefix

File: nanoc/test/checking/checks/test_internal_links.rb [hunk 3]
```diff
@@ -111,8 +125,22 @@ def test_unescape_url
...
       check = Nanoc::Checking::Checks::InternalLinks.create(site)
 
-      assert check.send(:valid?, 'stuff/right%20foo', 'output/origin')
-      refute check.send(:valid?, 'stuff/wrong%20foo', 'output/origin')
+      assert check.send(:valid?, path_to_file_uri('stuff/right%20foo', site), 'output/origin')
+      refute check.send(:valid?, path_to_file_uri('stuff/wrong%20foo', site), 'output/origin')
+    end
+  end
+
+  def deeply_nesten_relatives
```
Comment (ddfreyne) [inline nanoc/test/checking/checks/test_internal_links.rb]: Same here: add `test_` prefix

(I eventually want to move fully to RSpec, but it’s a slow work in progress.)

**Comment by da2x** [comment]
OK, one more go at this. test_nanoc_cruby24 and test_nanoc_cruby25 still not giving me any usable information.

**Comment by ddfreyne** [comment]
Whoa, `test_nanoc_cruby24` and `test_nanoc_cruby25` don’t seem to do anything… but not consistently. I’ve rescheduled test runs because it seems that _sometimes_, they do work. I haven’t been able to figure out why, and this is the first time I notice it.

**Comment by da2x** [comment]
No go. Try again?

**Comment by ddfreyne** [comment]
Very odd — it seems that this particular branch fails consistently, but only on CI. I can’t make it fail locally.

I created a branch to test out this weird behavior (#1433), but that branch also does not fail, and I can’t make it fail even when re-running over and over.

**Comment by da2x** [comment]
Lets run this branch with trace then.

TODO: Don’t merge without reverting https://github.com/nanoc/nanoc/pull/1422/commits/05c5d193841bbdb2a9c54c14bbd0f79b7346c40c

**Comment by da2x** [comment]
OK, so that gave us no new information.

**Comment by da2x** [comment]
I rebased this branch again to no avail. I managed to setup a Ruby 2.5 environment locally and the tests ran just fine.

@ddfreyne Any idea why the #1424 branch failed in the same way? There is really no reason why that branch should have fail. Could you try restarting the tests in that branch for comparison?

Could it be that CCI just doesn’t like branches created by me? :wink:

**Comment by ddfreyne** [comment]
I’m also stumped. I’ve never seen this behavior — I cannot figure out why `rake` would silently exit.

**Comment by ddfreyne** [comment]
I SSH’d into the test container (which you can do with CircleCI), changed the test configuration to be verbose, and got this:

```
$ bundle exec rake test --verbose --trace
** Invoke test (first_time)
** Invoke spec (first_time)
** Execute spec
/usr/local/bin/ruby -I/home/circleci/project/vendor/bundle/gems/rspec-core-3.8.0/lib:/home/circleci/project/vendor/bundle/gems/rspec-support-3.8.0/lib /home/circleci/project/vendor/bundle/gems/rspec-core-3.8.0/exe/rspec --pattern spec/\*\*\{,/\*/\*\*\}/\*_spec.rb

  1) Nanoc::Checking::Checks::ExternalLinks found has no issues
     Failure/Error: base_uri = URI::File.build([nil, filename])

     NameError:
       uninitialized constant URI::File
     # ./lib/nanoc/extra/link_collector.rb:88:in `uris_in_file'
     # ./lib/nanoc/extra/link_collector.rb:64:in `hrefs_in_file'
     # ./lib/nanoc/extra/link_collector.rb:43:in `block in filenames_per_href'
     # ./lib/nanoc/extra/link_collector.rb:78:in `block in grouped_filenames'
     # ./lib/nanoc/extra/link_collector.rb:77:in `each'
     # ./lib/nanoc/extra/link_collector.rb:77:in `grouped_filenames'
     # ./lib/nanoc/extra/link_collector.rb:43:in `filenames_per_href'
     # ./lib/nanoc/checking/checks/external_links.rb:14:in `run'
     # ./spec/nanoc/checking/checks/external_links_spec.rb:41:in `block (3 levels) in <top (required)>'
     # /home/circleci/project/common/spec/spec_helper_foot.rb:40:in `block (2 levels) in <top (required)>'
     # /home/circleci/project/common/spec/spec_helper_foot_core.rb:31:in `block (4 levels) in <top (required)>'
     # /home/circleci/project/common/spec/spec_helper_foot_core.rb:6:in `__nanoc_core_chdir'
     # /home/circleci/project/common/spec/spec_helper_foot_core.rb:31:in `block (3 levels) in <top (required)>'
     # /home/circleci/project/common/spec/spec_helper_foot_core.rb:30:in `block (2 levels) in <top (required)>'
```

… and a handful of similar failures.

* * *

FYI, I changed the `:spec` task in Rakefile to say

```
RSpec::Core::RakeTask.new(:spec) do |t|
  t.verbose = true
end
```

**Comment by ddfreyne** [comment]
`uninitialized constant URI::File` does not make sense to me, because Nanoc does require `uri`, and then `URI::File` becomes available…

**Comment by ddfreyne** [comment]
@da2x Ohhhh, `URI::File` does not exist in Ruby 2.5.x — it’s new in Ruby 2.6. That would explain…

**Comment by ddfreyne** [comment]
The “test summary” is supposed to list all the failed tests, but that is clearly not quite working the way I expected…

**Comment by da2x** [comment]
Thanks, trying URI::Generic instead. Seems to handle Windows paths as well so should be fine.

What about test_nanoc_cruby24?

**Comment by ddfreyne** [comment]
Can you apply the changes from #1434 to your branch? That should get you actual test results!

**Comment by da2x** [comment]
All green!

…

So there is a small behavioral difference in Ruby2.4 compared to 2.5 and 2.6 that I’ll just leave in. `URI::Generic` isn’t aware of the file URI scheme and blindly merges the protocol-level slashes. So it returns `file:/X` instead of `file:///X`  and `file:/c:/X` instead of `file://c:/X` on Windows. Everything works, notably the improved handling of deeply nested relative paths, but you’ll need to be aware of the difference when handling raw output from the LinkCollector. I’ve changed one test to account for that difference.

**Pull request merged**